src/containers/CustomFunctionModal.tsx

```diff
@@ -0,0 +1,212 @@
+import React, { Component, createRef } from 'react'
+import { connect } from 'react-redux'
+import { t } from 'ttag'
+import { variables } from '../config'
+import ModalCard from '../components/ModalCard'
+import {
+  createFunction as createFunctionConnect,
+  deleteFunction as deleteFunctionConnect,
+  setFunction as setFunctionConnect,
+} from '../actions/customFunctions'
+import parser, { getNames } from '../parser'
+import type { CustomFunction } from '../reducers/customFunctions'
+
+interface CustomFunctionModalProps {
+  customFunction: CustomFunction
+  deactivateModal: () => void
+  createFunction: (name: string, func: string) => void
+  setFunction: (id: string, name: string, func: string) => void
+  deleteFunction: (id: string) => void
+}
+
+interface CustomFunctionModalState {
+  name: string
+  func: string
+  nameError: string
+  funcError: string
+}
+
+class CustomFunctionModal extends Component<CustomFunctionModalProps, CustomFunctionModalState> {
+  nameRef: React.RefObject<HTMLInputElement>
+
+  funcRef: React.RefObject<HTMLTextAreaElement>
+
+  constructor(props: CustomFunctionModalProps) {
+    super(props)
+    const { customFunction: cf } = this.props
+
+    this.nameRef = createRef()
+    this.funcRef = createRef()
+    this.state = {
+      name: cf ? cf.name : '',
+      func: cf ? cf.func : '',
+      nameError: '',
+      funcError: '',
+    }
```

Comment:
I think `getDerivedStateFromProps` is when you update your state based on prop changes. This state is only _initialized_ from prop, but does not update based on prop changes.

Reply:
But _shouldn't_ it update whenever the prop changes?

Reply:
I don't think so... I'm managing form state (function and name) internally until you save it with the Save button or discard it with Cancel. I don't think we need to save anything to redux (which is how props would update) until that point. If we did, it could get clunky and complicated (async comes to mind).

Reply:
What I mean is: let's say usage of this changes down the road. Maybe we just hide the modal instead of conditionally rendering it, and so we un-hide it with new props. This would fail in that case no? Maybe not a super likely scenario, but on the other hand, we lose nothing by putting this in `getDerivedStateFromProps`: it'll work now and it'll work in any other use case where props change after construction.

Reply:
I can see that... but since it's not happening now I think it's better not to have it. Personally if I'm looking through the component later and I see that it uses `getDerivedStateFromProps`, I would assume that it might get updates from redux beyond when it is a hidden modal. Operating on that assumption makes jumping into the code more difficult.

Reply:
I guess what I'm saying is adding `getDerivedStateFromProps` makes me assume that we're using it, so it's a little confusing if we're not.

Reply:
On the other hand, maybe I'm _misusing_ the `ModalCard` component and not hiding it is confusing or inefficient.

Reply:
To be clear, I'm not saying we should be hiding/showing it. I think conditional rendering is the way to go. That was just to illustrate that `getDerivedStateFromProps` has an advantage over settings state from props in the constructor, and has no disadvantage to it.

Reply:
> `getDerivedStateFromProps` has an advantage over settings state from props in the constructor, and has no disadvantage to it

One disadvantage is that `getDerivedStateFromProps` runs on every render, instead of just on initialization. 

in the [documentation](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops): 

> This method exists for rare use cases where the state depends on changes in props over time [ours currently doesn't] ... Deriving state leads to verbose code and makes your components difficult to think about.

A bigger disadvantage is confusion. Using `getDerivedStateFromProps` in this component means state isn't limited to the component itself, it can get updated from across the app. In some areas of AMMP, field values are passed to redux on blur to avoid users from losing things as they move between views. If I were revisiting this component and saw `getDerivedStateFromProps`, I'd assume something like that was going on, or possibly updates could trickle in from loading a Save URL, or who knows why that code is there? (but I'd think certainly it would be there because we were using it, not for our potential to use it). I'd search app-wide for how `customFunctions` could update, and be confused that I couldn't find updates to `customFunctions` that update `name` or `func` outside of this component.

Reply:
Good points. This should still go in a lifecycle method though. `componentDidMount` seems appropriate.

Reply:
Also, I don't believe redux comes into play here, since it's not providing that prop.

Reply:
> This should still go in a lifecycle method though. componentDidMount seems appropriate.

Sounds good, though I'm not sure what the aversion to initializing state from a prop in the constructor is (why trigger a second update to the component immediately after mount?)

> I don't believe redux comes into play here, since it's not providing that prop.

True the prop comes directly from the parent, but the parent's parent gets `customFunctions` from redux.

Reply:
> I'm not sure why we want to avoid initializing state from a prop in the constructor (why trigger a second update to the component immediately after mount?)

Just that object construction and React component _mount_ are not necessarily the same thing. Sure, it seems to work (in that props are available when the constructor is called) but I'm not sure that's always _guaranteed_ to happen. E.g., I could imagine a scenario in which React creates an object from the class before taking props into account and _then_ mounts the component, invoking `componentDidMount`. Very unlikely, I'm sure, but a lifecycle method seems like the better place to put it.